### ISSUE TITLE
Expected schema checks not working on MariaDB 10.5

### ISSUE THREAD

**Issue by Gudmo** [MISP/MISP#6277]
Table name | Description | Expected schema | Actual schema |  -- | -- | -- | -- | --
attributes       Column `object_relation` is different | object_relation YES varchar 255 utf8_bin varchar(255) | object_relation YES varchar 255 utf8_bin varchar(255) NULL |  
                       Column `comment` is different | comment YES text 65535 utf8_unicode_ci text | comment YES text 65535 utf8_unicode_ci ext NULL |  
                       Column `first_seen` is different | first_seen YES bigint 19 bigint(20) | first_seen YES bigint 19 bigint(20) NULL |  
                       Column `last_seen` is different | last_seen YES bigint 19 bigint(20) | last_seen YES bigint 19 bigint(20) NULL

+---------------------+--------------+------+-----+---------+----------------+
| Field               | Type         | Null | Key | Default | Extra          |
+---------------------+--------------+------+-----+---------+----------------+
| id                  | int(11)      | NO   | PRI | NULL    | auto_increment |
| event_id            | int(11)      | NO   | MUL | NULL    |                |
| category            | varchar(255) | NO   | MUL | NULL    |                |
| type                | varchar(100) | NO   | MUL | NULL    |                |
| value1              | text         | NO   | MUL | NULL    |                |
| value2              | text         | NO   | MUL | NULL    |                |
| to_ids              | tinyint(1)   | NO   |     | 1       |                |
| uuid                | varchar(40)  | NO   | MUL | NULL    |                |
| timestamp           | int(11)      | NO   |     | 0       |                |
| distribution        | tinyint(4)   | NO   |     | 0       |                |
| sharing_group_id    | int(11)      | NO   | MUL | NULL    |                |
| comment             | text         | YES  |     | NULL    |                |
| deleted             | tinyint(1)   | NO   | MUL | 0       |                |
| disable_correlation | tinyint(1)   | NO   |     | 0       |                |
| object_id           | int(11)      | NO   | MUL | 0       |                |
| object_relation     | varchar(255) | YES  | MUL | NULL    |                |
| first_seen          | bigint(20)   | YES  | MUL | NULL    |                |
| last_seen           | bigint(20)   | YES  | MUL | NULL    |                |
+---------------------+--------------+------+-----+---------+----------------+
18 rows in set (0.001 sec)

It may be garbled, but the gist of it is that now the column info is different and MISP thinks the columns are all wrong, when they are in fact perfectly fine.

**Comment by mokaddem** [comment]
Are these warning shown when the `Reveal benign deltas` checkbox is checked or do they appear in the default view?

**Comment by gitnorty** [comment]
We see similar behavior in our instances; when the 'Reveal benign deltas' is checked most of the items are still displayed (these rows are in pink).  From what I'm seeing the difference is mostly due to the integer length configured.   The rows (in white) which are hidden with this box checked all seem to be different based on the Unicode version in use.  Our system says utf8mb4_0900_ai_ci is used whereas the expected schema says we should be using utf8mb4_general_ci...

The diagnostics page reports over 400 discrepancies.

Please let me know if more information is required?

Thanks!

**Comment by github-germ** [comment]
Related to:
* #6284 `Extend database schema diagnostic to ignore integer display-width specifications like "(11)" on int`

**Comment by Gudmo** [comment]
This occurs when Reveal benign deltas is checked, the same as with gitnorty's system.

**Comment by iglocska** [comment]
I think the whole point of it being a benign difference is that MISP does in fact not think that your db is wrong, just simply that there's a benign mismatch between the schema and the actual state (which in this case is fine).

**Comment by imidoriya** [comment]
Looking at all these, it's the `column default` for the field.  Many of these in MariaDB appear to default to `NULL`, where the MISP schema doesn't define a default, which I think is the same thing as `NULL`.  So I'd argue that they are the same and the check fails to correctly identify that they are the same.  From my tests, it's not something that ALTER table can fix.  Another field I see is that the MISP scheme will say that a varchar field should be empty, but the default is `''` or MISP defines the default as `network` and the value is `'network'`.  So I think a fix would be to accept Null for an empty value, strip the leading and trailing `'` in a char field when comparing.

<img width="691" alt="Screen Shot 2020-09-08 at 4 59 42 PM" src="https://user-images.githubusercontent.com/45568592/92527273-c3f2b980-f1f4-11ea-8b76-d5890711c4c5.png">

<img width="692" alt="Screen Shot 2020-09-08 at 4 59 55 PM" src="https://user-images.githubusercontent.com/45568592/92527301-c9500400-f1f4-11ea-9d3d-e48d9dbad617.png">

<img width="977" alt="Screen Shot 2020-09-08 at 4 53 04 PM" src="https://user-images.githubusercontent.com/45568592/92527194-a45b9100-f1f4-11ea-8812-27009c4a76e8.png">

**Comment by imidoriya** [comment]
To follow up on my last comment, I found the code that is responsible.  I'm not sure if it's intentional or not.  This appears to actually do the check, but IMO these distinctions are equivalent based on MySQL/MarinaDB, not benign errors.  It's not fixable with the SQL statement as that's just the way the DB does it.

https://github.com/MISP/MISP/blob/531e920b422a2907ba8d3e831b59a8e7957a4f9f/app/Model/Server.php#L2907-L2912

In this situation, would it make sense to just make them equal, instead of a benign error.  It's a distinction without a difference (likely just the way it reports it) and it shouldn't produce an error for the default value.
```
$actualValue = $expectedValue
```

**Comment by iglocska** [comment]
Yeah most of the benign errors are really just flavour differences between the DBs. We basically still show them in case we didn't catch something relevant and we ourselves need to debug it, but for all intents and purposes we ask every user of the system to just ignore benign errors altogether. Maybe we should just hide them completely.

The idea to make them equal would indeed be another option!

**Comment by imidoriya** [comment]
Yeah, it's an OCD thing.. I can't look at something that reports to be an error and not want to fix it. haha  The little alert icons on the Server Settings & Maintenance bar drive me crazy too. Yes, I know I've disabled email and correlation, it's intentional not a critical error! lol
